### DISCUSSION TITLE
MIMXRT1042 FlexCAN2 not working

### DISCUSSION THREAD

**Discussion by rtborg** [zephyrproject-rtos/zephyr#67579] [General] [answered]
I have a custom board with MIMXRT1042XJM5B, and cannot get FlexCAN2 working. I am able to configure and start the CAN device, however when I send a frame I get a error -114 (-ENETUNREACH, bus-off state). 
There's a CAN dongle on the CAN bus, and both nodes have termination resistors; bitrate is 250000.  

Edit: Observing CAN TX pin with a logic analyzer shows activity, but all it does is transmit garbage. 

Here are the relevant parts from zephyr.dts: 
```

chosen {
	zephyr,entropy = &trng;
	die-temp0 = &tempmon;
	zephyr,sram = &sdram0;
	zephyr,itcm = &itcm;
	zephyr,dtcm = &dtcm;
	zephyr,console = &lpuart1;
	zephyr,shell-uart = &lpuart1;
	zephyr,flash = &w25q64jvssiq;
	zephyr,flash-controller = &w25q64jvssiq;
	zephyr,code-partition = &slot0_partition;
	zephyr,canbus = &flexcan2;
};

/*******************/

flexcan2: can@401d4000 {
	compatible = "nxp,flexcan";
	reg = < 0x401d4000 0x1000 >;
	interrupts = < 0x25 0x0 >;
	interrupt-names = "common";
	clocks = < &ccm 0x900 0x68 0x12 >;
	clk-source = < 0x2 >;
	sample-point = < 0x36b >;
	status = "okay";
	bus-speed = < 0x3d090 >;
	pinctrl-0 = < &pinmux_flexcan2 >;
	pinctrl-names = "default";
	can-transceiver {
		max-bitrate = < 0x7a120 >;
	};
};

pinmux_flexcan2: pinmux_flexcan2 {
	phandle = < 0xff >;
	group0 {
		pinmux = < &iomuxc_gpio_b1_08_flexcan2_tx >, < &iomuxc_gpio_b1_09_flexcan2_rx >;
		drive-strength = "r0-6";
		slew-rate = "slow";
		nxp,speed = "100-mhz";
	};
};
```

Generated zephyr.dts file:
[zephyr-dts.txt](https://github.com/zephyrproject-rtos/zephyr/files/13930816/zephyr-dts.txt)

**Comment by DerekSnell** [comment]
Hi @rtborg ,

Unfortunately, the FlexCAN has not yet been enabled on the RT1040 EVK.  The [supported features](https://docs.zephyrproject.org/latest/boards/arm/mimxrt1040_evk/doc/index.html#supported-features) for that board detail which features have been enabled and tested.  The [MIMXRT1060-EVK](https://docs.zephyrproject.org/latest/boards/arm/mimxrt1064_evk/doc/index.html#mimxrt1060-evk) board is very similar, supports the FlexCAN, and would be the closest board to use as a reference.  Did you already refer to that board to confirm how it enables the FlexCAN?

Seeing data transmitted seems like a good sign, like you are close to the correct configuration.  Have you already confirmed the bitrate of the transmission is as you expect?  That would help confirm the clocking and baud rate are configured correctly.

Let us know what you find.

**Comment by rtborg** [comment, marked as answer]
I managed to get FlexCAN2 working; the problem was not in software but in using the wrong TVS for the CAN lines on the PCB.
